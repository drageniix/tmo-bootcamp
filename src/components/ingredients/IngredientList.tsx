import React, { useContext } from "react";
import { AppContext } from "../../context/AppContext";

export const IngredientList = (): React.ReactElement => {
  const { ingredients, onRemoveIngredient } = useContext(AppContext);

  return (
    <section data-testid="ingredients">
      <ul className="ingredientList">
        {ingredients.map((ingredient) => (
          <li className="ingredientList__item" key={ingredient.id}>
            <button onClick={onRemoveIngredient(ingredient.id)}>-</button>
            <img
              className="ingredientList__item--img"
              src={`https://spoonacular.com/cdn/ingredients_100x100/${ingredient.image}`}
            />
            <h1 className="ingredientList__item--title">{ingredient.name}</h1>
          </li>
        ))}
      </ul>
    </section>
  );
};

export default IngredientList;
