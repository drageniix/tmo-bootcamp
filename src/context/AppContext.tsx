import React, { createContext } from "react";
import { BasicIngredient, Recipe } from "../api/recipe";
import { useRecipeSearch } from "../hooks/context";

export type AppContextType = {
  onSearchIngredients?: (input: string) => Promise<void>;
  onAddIngredient?: (chosenIngredient: BasicIngredient) => Promise<void>;
  onRemoveIngredient?: (
    id: number
  ) => (e: React.MouseEvent<HTMLButtonElement, MouseEvent>) => Promise<void>;
  ingredients: BasicIngredient[];
  searchedIngredients: BasicIngredient[];
  recipes: Recipe[];
};

export const defaultContext: AppContextType = {
  ingredients: [],
  recipes: [],
  searchedIngredients: [],
};

export const AppContext = createContext(defaultContext);

export const AppProvider = ({
  children,
}: {
  children: JSX.Element;
}): JSX.Element => {
  const appProps = useRecipeSearch();

  return <AppContext.Provider value={appProps}>{children}</AppContext.Provider>;
};
