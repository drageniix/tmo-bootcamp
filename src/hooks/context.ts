import { useState } from "react";
import {
  BasicIngredient,
  searchIngredients,
  searchRecipes,
} from "../api/recipe";
import { AppContextType, defaultContext } from "../context/AppContext";

export const useRecipeSearch = (): AppContextType => {
  const [recipes, setRecipes] = useState(defaultContext.recipes);
  const [ingredients, setIngredients] = useState(defaultContext.ingredients);
  const [searchedIngredients, setSearchedIngredients] = useState(
    defaultContext.searchedIngredients
  );

  const onSearchIngredients = async (input: string) => {
    const ingredients = await searchIngredients(input);
    setSearchedIngredients(ingredients);
  };

  const onAddIngredient = async (chosenIngredient: BasicIngredient) => {
    if (!ingredients.includes(chosenIngredient)) {
      const updatedIngredients = [...ingredients, chosenIngredient];
      const recipes = await searchRecipes(
        updatedIngredients.map((ingredient) => ingredient.name)
      );

      setIngredients(updatedIngredients);
      setRecipes(recipes);
    }
  };

  const onRemoveIngredient = (id: number) => async (
    e: React.MouseEvent<HTMLButtonElement, MouseEvent>
  ) => {
    e.preventDefault();
    const updatedIngredients = ingredients.filter(
      (ingredient) => ingredient.id !== id
    );
    const recipes = await searchRecipes(
      updatedIngredients.map((ingredient) => ingredient.name)
    );

    setIngredients(updatedIngredients);
    setRecipes(recipes);
  };

  return {
    onSearchIngredients,
    onAddIngredient,
    onRemoveIngredient,
    searchedIngredients,
    ingredients,
    recipes,
  };
};
