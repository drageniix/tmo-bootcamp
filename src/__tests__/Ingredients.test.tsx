import React from "react";
import { cleanup, render, fireEvent } from "@testing-library/react";
import { AppContext, AppContextType } from "../context/AppContext";
import IngredientList from "../components/ingredients/IngredientList";
import mockIngredients from "./mocks/ingredients.json";

const renderForm = (context?: AppContextType) =>
  render(
    <AppContext.Provider value={context}>
      <IngredientList />
    </AppContext.Provider>
  );

describe("should test ingredient list", () => {
  afterEach(cleanup);

  it("should match list snapshot", () => {
    expect.assertions(3);

    const { getByTestId } = renderForm({
      ingredients: [mockIngredients.results[0], mockIngredients.results[1]],
      recipes: [],
      searchedIngredients: [],
    });

    const listComponent = getByTestId("ingredients") as HTMLInputElement;
    expect(listComponent).toMatchSnapshot();
    expect(listComponent.textContent).toContain(
      mockIngredients.results[0].name
    );
    expect(listComponent.textContent).toContain(
      mockIngredients.results[1].name
    );
  });

  it("should remove recipe when button is clicked", () => {
    expect.assertions(1);
    const mockedRemoveIngredient = jest.fn();

    const { getByRole } = renderForm({
      onRemoveIngredient: mockedRemoveIngredient,
      ingredients: [mockIngredients.results[1]],
      recipes: [],
      searchedIngredients: [],
    });

    const buttonComponent = getByRole("button") as HTMLInputElement;
    fireEvent.click(buttonComponent);

    expect(mockedRemoveIngredient).toHaveBeenCalledWith(
      mockIngredients.results[1].id
    );
  });
});
