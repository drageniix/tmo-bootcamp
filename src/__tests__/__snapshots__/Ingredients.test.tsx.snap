// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test ingredient list should match list snapshot 1`] = `
<section
  data-testid="ingredients"
>
  <ul
    class="ingredientList"
  >
    <li
      class="ingredientList__item"
    >
      <button>
        -
      </button>
      <img
        class="ingredientList__item--img"
        src="https://spoonacular.com/cdn/ingredients_100x100/banana-chips.jpg"
      />
      <h1
        class="ingredientList__item--title"
      >
        banana chips
      </h1>
    </li>
    <li
      class="ingredientList__item"
    >
      <button>
        -
      </button>
      <img
        class="ingredientList__item--img"
        src="https://spoonacular.com/cdn/ingredients_100x100/limoncello.jpg"
      />
      <h1
        class="ingredientList__item--title"
      >
        banana liqueur
      </h1>
    </li>
  </ul>
</section>
`;
